includes/Modules/Reader_Revenue_Manager/Synchronize_OnboardingState.php

```diff
@@ -0,0 +1,147 @@
+ *
+ * @since n.e.x.t
+ * @package   Google\Site_Kit\Modules\Reader_Revenue_Manager
+ */
+
+namespace Google\Site_Kit\Modules\Reader_Revenue_Manager;
+
+use Google\Site_Kit\Core\Permissions\Permissions;
+use Google\Site_Kit\Core\Storage\User_Options;
+use Google\Site_Kit\Modules\Reader_Revenue_Manager;
+
+/**
+ * Class for synchronizing the onboarding state.
+ */
+class Synchronize_OnboardingState {
+	/**
+	 * Cron event name for synchronizing the onboarding state.
+	 */
+	const CRON_SYNCHRONIZE_ONBOARDING_STATE = 'googlesitekit_cron_synchronize_onboarding_state';
+
+	/**
+	 * Reader_Revenue_Manager instance.
+	 *
+	 * @var Reader_Revenue_Manager
+	 */
+	protected $reader_revenue_manager;
+
+	/**
+	 * User_Options instance.
+	 *
+	 * @var User_Options
+	 */
+	protected $user_options;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param Reader_Revenue_Manager $reader_revenue_manager Reader Revenue Manager instance.
+	 * @param User_Options           $user_options           User_Options instance.
+	 */
+	public function __construct( Reader_Revenue_Manager $reader_revenue_manager, User_Options $user_options ) {
+		$this->reader_revenue_manager = $reader_revenue_manager;
+		$this->user_options           = $user_options;
+	}
+
+	/**
+	 * Registers functionality through WordPress hooks.
+	 */
+	public function register() {
+		add_action(
+			self::CRON_SYNCHRONIZE_ONBOARDING_STATE,
+			function () {
+				$this->synchronize_publication_data();
+			}
+		);
+	}
+
+	/**
+	 * Synchronizes the onboarding state.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return void
+	 */
+	protected function synchronize_publication_data() {
+		$owner_id     = $this->reader_revenue_manager->get_owner_id();
+		$restore_user = $this->user_options->switch_user( $owner_id );
+
+		if ( user_can( $owner_id, Permissions::VIEW_AUTHENTICATED_DASHBOARD ) ) {
+			$this->synchronize_onboarding_state();
+		}
+
+		$restore_user();
+	}
+
+	/**
+	 * Synchronizes the onboarding state.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return void
+	 */
+	protected function synchronize_onboarding_state() {
+		$connected = $this->reader_revenue_manager->is_connected();
+
+		// If not connected, return early.
+		if ( ! $connected ) {
+			return;
+		}
+
+		$settings         = $this->reader_revenue_manager->get_settings()->get();
+		$publication_id   = $settings['publicationID'];
+		$onboarding_state = $settings['publicationOnboardingState'];
+
+		if ( $connected && 'ONBOARDING_COMPLETE' !== $onboarding_state ) {
+			$publications = $this->reader_revenue_manager->get_data( 'publications' );
+
+			// If publications is empty, return early.
+			if ( empty( $publications ) ) {
+				return;
+			}
+
+			$publication = array_filter(
+				$publications,
+				function ( $pub ) use ( $publication_id ) {
+					return $pub->getPublicationId() === $publication_id;
+				}
+			);
+
+			// If publication is empty, return early.
+			if ( empty( $publication ) ) {
+				return;
+			}
+
+			$publication = reset( $publication );
```

Comment:
I think we should keep it as is considering the edge case where we couldn't find the matching publication. In that case accessing [0] may lead to the warning.

Reply:
I understand your concern. I feel initialising the variable `$publication` with an array is a little odd, as it is a singular term but holds an array. What if we change the name of the variable to `$filtered_publications`? Also, what do you think about using `[0]` to access the first element of the array instead of using `reset`, as I think that makes the code further readable and doesn't bother with the internal pointer.

```suggestion
			$filtered_publications = array_filter(
				$publications,
				function ( $pub ) use ( $publication_id ) {
					return $pub->getPublicationId() === $publication_id;
				}
			);

			// If there are no filtered publications, return early.
			if ( empty( $filtered_publications ) ) {
				return;
			}

			$publication = $filtered_publications[0];

```